python_modules/libraries/dagster-dbt/dagster_dbt/core/resources_v2.py

```diff
@@ -255,78 +255,90 @@ def to_default_asset_events(
             attached_node_unique_id = test_resource_props.get("attached_node")
-            is_generic_test = bool(attached_node_unique_id)
 
-            if (
-                dagster_dbt_translator.settings.enable_asset_checks
-                and has_asset_def
-                and is_generic_test  # singular tests aren't ingested as asset checks yet
-            ):
-                attached_node_resource_props: Dict[str, Any] = manifest["nodes"].get(
+            # If the test is singular, infer the attached node from the upstream nodes.
+            if len(upstream_unique_ids) == 1:
+                [attached_node_unique_id] = upstream_unique_ids
+
+            # If the test is singular, but has multiple dependencies, infer the attached node from
+            # from the dbt meta.
+            has_asset_key_config = bool(
+                (
+                    test_resource_props.get("config", {}).get("meta", {})
+                    or test_resource_props.get("meta", {})
+                )
+                .get("dagster", {})
+                .get("asset_key", [])
+            )
+
+            attached_node_dbt_resource_props = None
+            if attached_node_unique_id:
+                attached_node_dbt_resource_props = manifest["nodes"].get(
                     attached_node_unique_id
                 ) or manifest["sources"].get(attached_node_unique_id)
-                attached_node_asset_key = dagster_dbt_translator.get_asset_key(
-                    attached_node_resource_props
-                )
-                check_name = event_node_info["node_name"]
-
-                if (
-                    context
-                    and AssetCheckKey(asset_key=attached_node_asset_key, name=check_name)
-                    in context.selected_asset_check_keys
-                ):
-                    is_test_successful = node_status == TestStatus.Pass
-                    severity = (
-                        AssetCheckSeverity.WARN
-                        if node_status == TestStatus.Warn
-                        else AssetCheckSeverity.ERROR
-                    )
+            elif has_asset_key_config:
+                attached_node_dbt_resource_props = test_resource_props
 
-                    yield AssetCheckResult(
-                        passed=is_test_successful,
-                        asset_key=attached_node_asset_key,
-                        check_name=check_name,
-                        metadata=metadata,
-                        severity=severity,
-                    )
+            attached_node_asset_key = (
+                dagster_dbt_translator.get_asset_key(attached_node_dbt_resource_props)
+                if attached_node_dbt_resource_props
+                else None
+            )
+            check_name = test_resource_props["name"]
+
+            # If the test was not selected as an asset check, yield an `AssetObservation`.
+            if not (
```

Comment:
I inverted the conditional logic of the function so that we could consolidate to a single `_yield_observation_events_for_test` call.